clang/lib/StaticAnalyzer/Checkers/BuiltinFunctionChecker.cpp

```diff
@@ -50,6 +101,44 @@ class BuiltinFunctionChecker : public Checker<eval::Call> {
 
 } // namespace
 
+void BuiltinFunctionChecker::HandleOverflowBuiltin(const CallEvent &Call,
+                                                   CheckerContext &C,
+                                                   BinaryOperator::Opcode Op,
+                                                   QualType ResultType) const {
+  // All __builtin_*_overflow functions take 3 argumets.
+  assert(Call.getNumArgs() == 3);
```

Comment:
The use of `assert()` should be limited to situations where we realize that the _analyzer code_ is buggy (we reached an "impossible" situation). If the _analyzed code_ contains an invalid call like `__builtin_add_overflow(10, 20, &res, "spam")`, then the analyzer _may_ report an error, but must not crash with an assertion failure.

Unfortunately, this particular checker is a so-called "modeling" checker, so it is hidden from the user (as an undocumented implementation detail), and therefore it cannot create bug reports.

This means that if we encounter an invalid `__builtin_*_overflow` call, then we should probably just ignore it, because we should not assert and we cannot create a bug report. I'd assume that this is an extremely rare situation (if someone uses these builtin function, they're unlikely to mess up the argument count), so a more complex solution (e.g. introducing a new non-modeling checker which creates bug reports) is probably not worth the effort.

Reply:
Hm, I thought that CSA won't run on code with compile errors. Or there is a mode which will try to run even on invalid AST?

I've tried to catch this assert locally by calling `__builtin_*_overflow`  with 2 arguments, but clang just rejects it.

Reply:
I'm okay having that assert. We are talking about a builtin here.
I would have a different opinion if it wasn't a builtin though.

Reply:
So.... I've pushed what I came up with for handling overflow, but during test writing I found smth I don't understand. I've decided to push current state, since it's easier to show code than describe it =)

My current problem is following code:
```c
   if (a > 10)
     return;
   if (b > 10)
     return;

   // clang_analyzer_eval(a + b < 30); <--- Prints 1 and 0, but why ???
````
For some reason constraints do not work as expected. And because of that my overflow checker splits state where it shouldn't.... I'd really appreciate tips

Reply:
> I've tried to catch this assert locally by calling __builtin_*_overflow with 2 arguments, but clang just rejects it.

Oh, I see. Perhaps add a comment saying that "Calling a builtin with an incorrect argument count produces compiler error (...)", because e.g. I didn't know that this is the case.

Reply:
Added comments to the asserts (about number of arguments and result type) and added taint propagation